doc/decisions/DR015-Strategy-for-methods-with-several-signature-variations.md

```diff
@@ -0,0 +1,179 @@
+> the specifics of any particular signature/calling convention.
+
+## Relevant data
+
+- OpenAssetIO supports multiple languages (currently C, C++, Python),
+  each with their own syntax and language rules. It will certainly be
+  extended in the future to as yet unknown languages with similarly
+  unknown constraints.
+
+- OpenAssetIO strives for coherence between languages as many developers
+  have to work in multi-language code bases. There are certainly
+  concessions if this leads to idiosyncrasies, but ideally terminology
+  and semantic structure is coherent wherever possible.
+
+The examples illustrated in the options below focus on the case of
+providing "host conveniences" for the batch-centric API methods. The
+OpenAssetIO batch API uses the following canonical from:
+
+```c++
+void someSingularName(inputVector, ..., successCallback, errorCallback);
+```
+
+The rationale for this form is covered
+[here](./DR009-Batch-method-result-types.md) and
+[here](./DR003-ManagerInterface-method-naming.md). The salient points of
+the purposes of this discussion are:
+
+- There is no return value.
+- The `successCallback` is called for each element of the input vector
+  that is processed without error.
+- The `errorCallback` is called for any element of the input vector that
+  resulted in a processing error.
+
+Though this is flexible, and supports future design patterns, it can add
+significantly to the host implementation boilerplate for simple tasks
+such as resolving a single reference where any error is considered
+exceptional.
+
+We desire to add several host-facing conveniences that simplify common
+API interactions. This immediately precipitates the discussion as to how
+these should be integrated into the API.
+
+## Options considered
+
+### Option 1
+
+Use function overloading where possible.
+
+```c++
+void resolve(vector<T> inputs, F* successCallback, F* errorCallback);
+void resolve(vector<T> inputs, vector<variant<ValueType, Error>> &out);
+void resolve(vector<T> inputs, vector<ValueType> &out);
+void resolve(T input, variant<ValueType, Error> &out);
+void resolve(T input, ValueType &out);
+
+void entityVersions(vector<T> inputs, F* successCallback, F* errorCallback);
+void entityVersions(vector<T> inputs, vector<variant<ValueType, Error>> &out);
+void entityVersions(vector<T> inputs, vector<ValueType> &out);
+void entityVersions(T input, variant<ValueType, Error> &out);
+void entityVersions(T input, ValueType &out);
+```
+
+#### Pros
+
+- Low cognitive overhead to determine which high-level API method is
+  being invoked as the name contains no other information.
+- Avoids awkward compromises in determining method name mutations.
+
+#### Cons
+
+- Function overloading is not supported in many languages.
+- High cognitive overhead in understanding which variation is being used
+  as it relies on parameter naming and/or type visibility.
+- Forces out-params in C++ as overloading can't be based on return type.
+
```

Comment:
Con: Hides implicit conversions. For example, Imath::half has implicit conversions to and from float, float has implicit conversions to and from double. This can yield loss of information not reported by the compiler.

Reply:
Added note [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/824/commits/04dda917c3b4e8f30118c6f120d9a6b0ea972828)